Spigot-Server-Patches/0212-Use-TerminalConsoleAppender-for-console-improvements.patch

```diff
@@ -0,0 +1,659 @@
+             protected List<String> evaluate() {
+@@ -30,25 +42,39 @@ public class ConsoleCommandCompleter implements Completer {
+                 return tabEvent.isCancelled() ? Collections.EMPTY_LIST : tabEvent.getCompletions();
+             }
+         };
+-        this.server.getServer().processQueue.add(waitable);
++        server.getServer().processQueue.add(waitable); // Paper - Remove "this."
+         try {
+             List<String> offers = waitable.get();
+             if (offers == null) {
+-                return cursor;
++                return; // Paper - Method returns void
+             }
+-            candidates.addAll(offers);
+ 
++            // Paper start - JLine update
++            //candidates.addAll(offers);
++            for (String completion : offers) {
++                if (completion.isEmpty()) {
++                    continue;
++                }
++
++                candidates.add(new Candidate(completion));
++            }
++            // Paper end
++
++            // Paper start - JLine handles cursor now
++            /*
+             final int lastSpace = buffer.lastIndexOf(' ');
+             if (lastSpace == -1) {
+                 return cursor - buffer.length();
+             } else {
+                 return cursor - (buffer.length() - lastSpace - 1);
+             }
++            */
++            // Paper end
+         } catch (ExecutionException e) {
+-            this.server.getLogger().log(Level.WARNING, "Unhandled exception when tab completing", e);
++            server.getLogger().log(Level.WARNING, "Unhandled exception when tab completing", e); // Paper - Remove "this."
+         } catch (InterruptedException e) {
+             Thread.currentThread().interrupt();
+         }
+-        return cursor;
++        //return cursor; // Paper - Method returns void
+     }
+ }
+diff --git a/src/main/java/org/bukkit/craftbukkit/util/ServerShutdownThread.java b/src/main/java/org/bukkit/craftbukkit/util/ServerShutdownThread.java
+index ae3481ba..358e8e91 100644
+--- a/src/main/java/org/bukkit/craftbukkit/util/ServerShutdownThread.java
++++ b/src/main/java/org/bukkit/craftbukkit/util/ServerShutdownThread.java
+@@ -18,7 +18,10 @@ public class ServerShutdownThread extends Thread {
+             ex.printStackTrace();
+         } finally {
+             try {
+-                server.reader.getTerminal().restore();
++                // Paper start - Use TerminalConsoleAppender
++                //server.reader.getTerminal().restore();
++                net.minecrell.terminalconsole.TerminalConsoleAppender.close();
++                // Paper end
+             } catch (Exception e) {
+             }
+         }
+diff --git a/src/main/java/org/bukkit/craftbukkit/util/TerminalConsoleWriterThread.java b/src/main/java/org/bukkit/craftbukkit/util/TerminalConsoleWriterThread.java
+deleted file mode 100644
+index b6409711..00000000
+--- a/src/main/java/org/bukkit/craftbukkit/util/TerminalConsoleWriterThread.java
```

Comment:
same here, let's avoid deleting the files to keep diff/patch smaller and reduce impact of silly plugins doing silly things.